sql/distinct.go

```diff
@@ -0,0 +1,137 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// distinct constructs a distinctNode.
+func (*planner) distinct(n *parser.Select) *distinctNode {
+	if n.Distinct {
+		distinct := distinctNode{}
+		distinct.suffixSeen = make(map[string]struct{})
+		return &distinct
+	}
+	return nil
+}
+
+type distinctNode struct {
+	planNode
+	// All the columns that are part of the Sort. Set to nil if no-sort, or
+	// sort used an expression that was not part of the requested column set.
+	columnsInOrder []bool
+	// encoding of the columnsInOrder columns for the previous row.
+	prefixSeen string
+	// encoding of the non-columnInOrder columns for rows sharing the same
+	// prefixSeen value.
+	suffixSeen map[string]struct{}
+}
+
+// wrap the supplied planNode with the distinctNode if distinct is required.
+func (n *distinctNode) wrap(plan planNode) planNode {
+	if n == nil {
+		return plan
+	}
+	n.planNode = plan
+	if len(plan.Ordering()) > 0 {
+		n.columnsInOrder = make([]bool, len(plan.Columns()))
+	}
+	for _, p := range plan.Ordering() {
+		if p < 0 {
+			p = -p
+		}
+		if p <= len(n.columnsInOrder) {
+			n.columnsInOrder[p-1] = true
+		} else {
+			// Cannot use sort order for distinct.
+			n.columnsInOrder = nil
+			break
+		}
+	}
+	return n
+}
+
+func (n *distinctNode) Next() bool {
+	for n.planNode.Next() {
+		// Detect duplicates
+		prefix, suffix := n.encodeValues(n.Values())
+		pKey := string(prefix)
+		sKey := string(suffix)
+		if pKey == n.prefixSeen {
+			// The prefix of the row is the same as the last row; check
+			// to see if the suffix which is not ordered has been seen.
+			if sKey == "" {
+				// duplicate
+				continue
+			}
+
+			if _, ok := n.suffixSeen[sKey]; ok {
+				// duplicate
+				continue
+			}
+		} else {
+			// The prefix of the row which is ordered differs from the last row;
+			// reset our seen set.
+			if len(n.suffixSeen) > 0 {
+				n.suffixSeen = make(map[string]struct{})
+			}
+			n.prefixSeen = pKey
+		}
+		if sKey != "" {
+			n.suffixSeen[sKey] = struct{}{}
+		}
+		return true
+	}
+	return false
+}
+
+func (n *distinctNode) encodeValues(values parser.DTuple) ([]byte, []byte) {
+	prefix := make([]byte, 0, 100)
+	suffix := make([]byte, 0, 100)
+	for i, val := range values {
+		var err error
+		if n.columnsInOrder != nil && n.columnsInOrder[i] {
+			prefix, err = encodeTableKey(prefix, val)
+		} else {
+			suffix, err = encodeTableKey(suffix, val)
+		}
+		if err != nil {
+			panic(fmt.Sprintf("unable to encode value: %v", val))
```

Comment:
Probably should return an error here. This is usually done by having a `distinctNode.err` field and implementing the `Err()` method. You'll also need to handle the case where `n.planNode.Next()` returns false because an error occurred.